tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/smt/Z3SolverContext.scala

```diff
@@ -218,45 +195,46 @@ class Z3SolverContext(val config: SolverConfig) extends SolverContext {
-              val name = setExpr.toString + "_" + s"in_${elemT.signature}${elemId}_${setT.signature}$setId"
-              logWriter.flush() // flush the SMT log
-              throw new IllegalStateException(s"SMT $id: The Array constant $name is missing from the SMT context")
-
-            // Array "store"
-            case Some(list) if list.head._1 == false =>
-              val storedExpr = if (propagatedArraySelect._1) propagatedArraySelect._2 else z3context.mkTrue()
-              propagatedArraySelect = propagatedArraySelect.copy(_1 = false)
-              val store = z3context.mkStore(
-                  setExpr.asInstanceOf[Expr[ArraySort[Sort, BoolSort]]], elemExpr.asInstanceOf[Expr[Sort]],
-                  storedExpr.asInstanceOf[Expr[BoolSort]])
-              val eqStore = z3context.mkEq(const, store)
-              // inStored records that the array was set in the current level
-              inStored += ((setId, elemId) -> ((true, level) :: inStored(setId, elemId)))
-              // cellCache records the current SSA array representing the set being handled
-              cellCache += (setId -> ((const, setT, level) :: cellCache(setId)))
-              eqStore.asInstanceOf[ExprSort]
-
-            // Array "select"
-            case Some(list) if list.head._1 == true =>
-              z3context
-                .mkSelect(setExpr.asInstanceOf[ArrayExpr[Sort, BoolSort]], elemExpr.asInstanceOf[Expr[Sort]])
-                .asInstanceOf[ExprSort]
-          }
-        } else {
-          const
-        }
+        const
     }
   }
 
+  private def mkSelect(setId: Int, elemId: Int): ExprSort = {
+    val set = cellCache(setId).head._1
+    val elem = cellCache(elemId).head._1
+
+    z3context
+      .mkSelect(set.asInstanceOf[ArrayExpr[Sort, BoolSort]], elem.asInstanceOf[Expr[Sort]])
+      .asInstanceOf[ExprSort]
+  }
+
+  private def mkStore(setId: Int, elemId: Int): ExprSort = {
+    val (set, setT, _) = cellCache(setId).head
+    val (elem, elemT, _) = cellCache(elemId).head
+
+    val edgeName = s"in_${elemT.signature}${elemId}_${setT.signature}$setId"
+    val updatedSetName = set.toString + "_" + edgeName
+    val setSort = getOrMkCellSort(setT)
+    log(s";; declare edge inclusion $edgeName")
+    log(s"(declare-const $updatedSetName $setSort)")
+    val updatedSet = z3context.mkConst(updatedSetName, setSort)
+    cellCache += (setId -> ((updatedSet, setT, level) :: cellCache(setId)))
+    _metrics = _metrics.addNConsts(1)
+
+    val store = z3context.mkStore(
+        set.asInstanceOf[Expr[ArraySort[Sort, BoolSort]]], elem.asInstanceOf[Expr[Sort]], z3context.mkTrue())
+    val eqStore = z3context.mkEq(updatedSet, store)
+    eqStore.asInstanceOf[ExprSort]
+  }
+
+  private def mkUnchanged(setId: Int): ExprSort = {
+    val currentset = cellCache(setId).head._1
+    val oldSet = if (cellCache(setId).size > 1) cellCache(setId).tail.head._1 else currentset
+
+    val eqUnchanged = z3context.mkEq(currentset, oldSet)
```

Comment:
I see that the behaviour of unchanged is not very clear. In the OOPSLA encoding set membership is undefined initially, and it needs to be set to true or false via `\in` or `\notin`. Since all arrays are initialized as the empty set, by default `\A x . x \notin Some_Set` and if membership wants to be asserted a new, updated, set needs to be defined, explicitly adding the element.

That is why (I believe) keeping the array as it is works here. I thought of adding a `not(select(...))` constraint, but I think it would be redundant. Every `\in` and `\notin` has a `read` and a `write` version, as you want to call them, and figuring out which version to use where was a good part of the problems I had with this PR. For `\notin`, its `read` version became `not(mkRead)`, while its `write` version became `mkUnchanged`, as it is already not in the set by default. A problem can arise if we substitute `\notin` by the wrong factory call, but if the rules are all correct, the approach should be sound.

Reply:
Regarding the trivial case, I noticed it but was not sure if I should simply assert true. Will make the change.

Reply:
I think we have to be very careful about "unchanged" when we are removing elements, not adding them. Maybe it is worth adding directly in the comment for unchanged.